pkg/cluster/builder.go

```diff
@@ -0,0 +1,119 @@
+package cluster
+
+import (
+	"github.com/pkg/errors"
+
+	"github.com/aws/eks-anywhere/pkg/constants"
+	"github.com/aws/eks-anywhere/pkg/manifests"
+	"github.com/aws/eks-anywhere/pkg/manifests/bundles"
+	"github.com/aws/eks-anywhere/pkg/version"
+	releasev1 "github.com/aws/eks-anywhere/release/api/v1alpha1"
+)
+
+// FileSpecBuilder allows to build [Spec] by reading from files.
+type FileSpecBuilder struct {
+	// TODO(g-gaston): this is very much a CLI thing. Move to `pkg/cli` when available.
+	reader              manifests.FileReader
+	cliVersion          version.Info
+	releasesManifestURL string
+	bundlesManifestURL  string
+}
+
+// FileSpecBuilderOpt allows to configure [FileSpecBuilder].
+type FileSpecBuilderOpt func(*FileSpecBuilder)
+
+// WithReleasesManifest configures the URL to read the Releases manifest.
+func WithReleasesManifest(url string) FileSpecBuilderOpt {
+	return func(b *FileSpecBuilder) {
+		b.releasesManifestURL = url
+	}
+}
+
+// WithOverrideBundlesManifest configures the URL to read the Bundles manifest.
+// This overrides the Bundles declared in the Releases so reading the Releases
+// manifest is skipped.
+func WithOverrideBundlesManifest(url string) FileSpecBuilderOpt {
+	return func(b *FileSpecBuilder) {
+		b.bundlesManifestURL = url
+	}
+}
+
+// NewFileSpecBuilder builds a new [FileSpecBuilder].
+// cliVersion is used to chose the right Bundles from the the Release manifest.
+func NewFileSpecBuilder(reader manifests.FileReader, cliVersion version.Info, opts ...FileSpecBuilderOpt) FileSpecBuilder {
```

Comment:
whats the reason to return `FileSpecBuilder` instead of `*FileSpecBuilder` here?

Reply:
It doesn't need to be a pointer because it doesn't need to modify itself after creation. Also, the amount of data it stores doesn't require passing it by reference, the copy shouldn't be a performance problem.

I usually go back to this when in doubt of what to use: https://github.com/golang/go/wiki/CodeReviewComments#receiver-type